pwiz_tools/Skyline/Test/CodeInspectionTest.cs

```diff
@@ -65,46 +71,141 @@ public void CodeInspection()
+
+                        if (requiredPatternsObservedInThisFile != null)
+                        {
+                            foreach (var requirement in requiredPatternsObservedInThisFile.Where(p => !p.Value))
                             {
-                                var why = forbidden[fileMask][pattern];
-                                result.Add(@"Code inspection failure: prohibited use of");
-                                result.Add(@"""" + pattern + @"""");
-                                result.Add("(" + why + ") at");
-                                result.Add(filename + "(" + lineNum + @"):");
-                                result.Add(line);
+                                var why = requiredPatternsByFileMask[fileMask][requirement.Key].Reason;
+                                result.Add(@"Did not find required use of");
+                                result.Add(@"""" + requirement.Key + @"""");
+                                result.Add("(" + why + ") in");
+                                result.Add(filename);
                                 result.Add(string.Empty);
                             }
                         }
-                        file.Close();
+
+                        if (result.Any())
+                        {
+                            results.Add(string.Join(Environment.NewLine, result));
+                        }
                     }
                 }
-                if (result.Any())
+
+                if (results.Any())
                 {
-                    AssertEx.Fail(string.Join(Environment.NewLine, result));
+                    results.Insert(0, string.Empty);
+                    results.Insert(0, string.Format("{0} code inspection failures found:", results.Count));
+                    AssertEx.Fail(string.Join(Environment.NewLine, results));
                 }
             }
         }
+
+
+        public class PatternDetails
+        {
+            public string Cue; // If non-empty, pattern only applies to files containing this cue
+            public string Reason; // Note to show on failure
+            public string[] IgnoredDirectories; // Don't flag on hits in these directories
+
+            public bool IgnorePath(string path)
+            {
+                return string.IsNullOrEmpty(path) ||
+                       IgnoredDirectories != null && IgnoredDirectories.Any(path.Contains);
+            }
+        }
+        private readonly Dictionary<string, Dictionary<string, PatternDetails>> forbiddenPatternsByFileMask = new Dictionary<string, Dictionary<string, PatternDetails>>();
+        private readonly Dictionary<string, Dictionary<string, PatternDetails>> requiredPatternsByFileMask = new Dictionary<string, Dictionary<string, PatternDetails>>();
+
+        private enum Inspection { Forbidden, Required }
+
+        private string[] NonLocalizedProjectDirectories = { "Skyline\\Executables", "Skyline\\TestUtil", "Skyline\\SkylineNightly" };
```

Comment:
The list of folders that we localize is in:
pwiz_tools\Skyline\Executables\KeepResx\Program.cs

It would be nice to have some way of keeping that folder list, and your folder list here in sync with each other.

Reply:
That's a handy list of things we do not localize, which I was unaware of. I'll see about using that.